### PR TITLE
Replace CAffectedKeysVisitor with descriptor based logic

### PR DESCRIPTION
**Pull request by sipa**
It seems we don't need a custom visitor pattern anymore to find what keys are affected by a script. Instead, infer the descriptor, and see which keys it expands to.

### CODE REVIEW

**Comment by instagibbs** [comment]
>+13 −67
concept ACK, note that this will now include pubkeys you have, not only private keys, which I think is a more sane general anyways.

**Comment by MeshCollider** [comment]
Concept ACK

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -104,67 +105,17 @@ std::string COutput::ToString() const
...
-        }
-    }
-
-    void operator()(const WitnessV0KeyHash& keyid)
-    {
-        CKeyID id(keyid);
-        if (keystore.HaveKey(id)) {
-            vKeys.push_back(id);
-        }
+std::vector<CKeyID> GetAffectedKeys(const CScript& spk, const SigningProvider& provider)
```
Comment (promag) [inline src/wallet/wallet.cpp]: nit, `static`. Could also keep some comment.

**Review by promag — CHANGES_REQUESTED** [review]
utACK 0e75f44.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: Could even make it more specific with the updated behavior.

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14821/commits/0e75f44a09ed07ff5eeeb142a4cefc4aba0ddf32

**Comment by instagibbs** [comment]
@sipa if you'd like you could take this test: https://github.com/instagibbs/bitcoin/commit/a80edd7e689e89e40d4ba98f3770c5da2bc8d174

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (Empact) [reply src/wallet/wallet.cpp]: nit: script_pub_key would be a more readable arg name. Granted I’m a noob but I had to check the callers.

**Comment by jonasschnelli** [comment]
utACK 0e75f44a09ed07ff5eeeb142a4cefc4aba0ddf32
also in favour to pull in 
https://github.com/instagibbs/bitcoin/commit/a80edd7e689e89e40d4ba98f3770c5da2bc8d174

**Comment by instagibbs** [comment]
@jonasschnelli I ended up PRing it, that commit is bad anyways: https://github.com/bitcoin/bitcoin/pull/14857

**Comment by sipa** [comment]
@instagibbs #14857 is merged, any suggested improvements to this PR left?

**Comment by instagibbs** [comment]
sorry! no

utACK https://github.com/bitcoin/bitcoin/pull/14821/commits/0e75f44a09ed07ff5eeeb142a4cefc4aba0ddf32

**Pull request merged**